src/ledger/LedgerManagerImpl.cpp

```diff
@@ -1573,6 +1576,430 @@ LedgerManagerImpl::prefetchTransactionData(AbstractLedgerTxnParent& ltx,
+            if (!it.second)
+            {
+                it.first->second = add_sat(it.first->second, delta.second);
+            }
+        }
+    }
+
+    LedgerTxn ltxInner(ltx);
+    for (auto const& thread : stage)
+    {
+        for (auto const& txBundle : thread)
+        {
+            // First check the invariants
+            if (txBundle.getResPayload()->isSuccess())
+            {
+                try
+                {
+                    // Soroban transactions don't have access to the ledger
+                    // header, so they can't modify it. Pass in the current
+                    // header as both current and previous.
+                    txBundle.getEffects().getDelta().header.current =
+                        ltxInner.loadHeader().current();
+                    txBundle.getEffects().getDelta().header.previous =
+                        ltxInner.loadHeader().current();
+                    app.checkOnOperationApply(
+                        txBundle.getTx()->getRawOperations().at(0),
+                        txBundle.getResPayload()->getOpResultAt(0),
+                        txBundle.getEffects().getDelta());
+                }
+                catch (InvariantDoesNotHold& e)
+                {
+                    printErrorAndAbort(
+                        "Invariant failure while applying operations: ",
+                        e.what());
+                }
+            }
+
+            txBundle.getTx()->processPostApply(mApp.getAppConnector(), ltxInner,
+                                               txBundle.getEffects().getMeta(),
+                                               txBundle.getResPayload());
+
+            // We only increase the internal-error metric count if the
+            // ledger is a newer version.
+            if (txBundle.getResPayload()->getResultCode() == txINTERNAL_ERROR &&
+                ledgerInfo.getLedgerVersion() >=
+                    config.LEDGER_PROTOCOL_MIN_VERSION_INTERNAL_ERROR_REPORT)
+            {
+                auto& internalErrorCounter = app.getMetrics().NewCounter(
+                    {"ledger", "transaction", "internal-error"});
+                internalErrorCounter.inc();
+            }
+        }
+    }
+
+    for (auto const& threadEntryMap : entryMapsByThread)
+    {
+        for (auto const& entry : threadEntryMap)
+        {
+            // Only update if dirty bit is set
+            if (!entry.second.isDirty)
+            {
+                continue;
+            }
+
+            if (entry.second.mLedgerEntry)
+            {
+                auto const& updatedEntry = *entry.second.mLedgerEntry;
+                auto ltxe = ltxInner.load(entry.first);
+                if (ltxe)
+                {
+                    if (ltxe.current().data.type() == TTL)
+                    {
+                        auto currLiveUntil =
+                            ltxe.current().data.ttl().liveUntilLedgerSeq;
+
+                        releaseAssertOrThrow(
```

Comment:
This assertion is not correct, is it? One can delete an entry in tx 1, and then re-create it with smaller TTL in tx 2. We only enforce non-decreasing TTL per transaction, but not across transactions. It's probably worth adding a respective test as well.

Reply:
Good catch. Fixed and added a test.